File: Project/MathParserManaged/Expression.cs
// Copyright 2019. All Rights Reserved.
using System.Text;
using Zorvan.Framework.MathParser.Syntax;
using Zorvan.Framework.MathParser.SyntaxTree;

namespace Zorvan.Framework.MathParser
{
	public class Expression
	{
		private TreeNode node = null;

		public Expression(string Expr)
		{
			node = new Lexer(Expr).Parse();
		}

		public double Calculate(params Argument[] Arguments)
		{
			ArgumentsMap arguments = new ArgumentsMap();
			if (Arguments != null)
				for (int i = 0; i < Arguments.Length; ++i)
				{
					Argument arg = Arguments[i];
					arguments[arg.Name] = arg.Value;
				}

			return TreeNodeCalculator.Calculate(node, arguments);
		}

		public override string ToString()
		{
			StringBuilder builder = new StringBuilder();

			Print(node, 0, builder);

			return builder.ToString();
		}

		private static void Print(TreeNode Node, int Indent, StringBuilder Builder)
		{
			++Indent;

			if (Node is NumberNode)
			{
				PrintLine(((NumberNode)Node).Number.ToString(), Indent, Builder);
			}
			else if (Node is ParameterNode)
			{
				PrintLine(((ParameterNode)Node).Name, Indent, Builder);
			}
			else if (Node is FunctionNode)
			{
				FunctionNode node = (FunctionNode)Node;

				PrintLine(node.Name + " (", Indent, Builder);
				TreeNodeCollection nodes = node.Parameters;
				for (int i = 0; i < nodes.Count; ++i)
					Print(nodes[i], Indent + 1, Builder);
				PrintLine(" )", Indent, Builder);
			}
			else if (Node is LogicalOperatorNode)
			{
				LogicalOperatorNode node = (LogicalOperatorNode)Node;

				Print(node.LeftNode, Indent, Builder);
				PrintLine(node.Operator.ToString(), Indent, Builder);
				Print(node.RightNode, Indent, Builder);
			}
			else if (Node is ArithmeticOperatorNode)
			{
				ArithmeticOperatorNode node = (ArithmeticOperatorNode)Node;

				Print(node.LeftNode, Indent, Builder);
				PrintLine(node.Operator.ToString(), Indent, Builder);
				Print(node.RightNode, Indent, Builder);
			}
			else
				throw new System.Exception("Unknown TreeNode type");
		}

		private static void PrintLine(string Text, int Indent, StringBuilder Builder)
		{
			for (int i = 0; i < Indent; ++i)
				Builder.Append(' ');

			Builder.AppendLine(Text);
		}
	}
}

File: Project/MathParserNative/Compile/Expression.cpp
// Copyright 2019. All Rights Reserved.
#include "..\Include\Expression.h"
#include "..\Include\Syntax\Lexer.h"
#include "..\Include\SyntaxTree\ArithmeticOperatorNode.h"
#include "..\Include\SyntaxTree\FunctionNode.h"
#include "..\Include\SyntaxTree\LogicalOperatorNode.h"
#include "..\Include\SyntaxTree\NumberNode.h"
#include "..\Include\SyntaxTree\ParameterNode.h"

using namespace Zorvan::Framework::MathParser::Syntax;

namespace Zorvan::Framework::MathParser
{
	Expression::Expression(const std::string & Expr)
	{
		m_Node = Lexer(Expr).Parse();
	}

	double Expression::Calculate(const ArgumentCollection &Arguments)
	{
		TreeNodeCalculator::ArgumentsMap arguments;

		for each (auto arg in Arguments)
			arguments[arg->GetName()] = arg->GetValue();

		return TreeNodeCalculator::Calculate(m_Node, arguments);
	}

	std::string Expression::ToString(void) const
	{
		std::stringstream builder;

		Print(m_Node, 0, builder);

		return builder.str();
	}

	void Expression::Print(TreeNode *Node, int Indent, std::stringstream &Builder)
	{
		++Indent;

		if (IS_CHILD_OF(Node, NumberNode))
		{
			std::stringstream ss;
			ss << reinterpret_cast<NumberNode*>(Node)->GetNumber();
			PrintLine(ss.str(), Indent, Builder);
		}
		else if (IS_CHILD_OF(Node, ParameterNode))
		{
			PrintLine(reinterpret_cast<ParameterNode*>(Node)->GetName(), Indent, Builder);
		}
		else if (IS_CHILD_OF(Node, FunctionNode))
		{
			FunctionNode *node = reinterpret_cast<FunctionNode*>(Node);

			PrintLine(node->GetName() + " (", Indent, Builder);

			const TreeNodeCollection &nodes = node->GetParameters();
			for each (const auto &node in nodes)
				Print(node, Indent + 1, Builder);

			PrintLine(" )", Indent, Builder);
		}
		else if (IS_CHILD_OF(Node, LogicalOperatorNode))
		{
			LogicalOperatorNode *node = reinterpret_cast<LogicalOperatorNode*>(Node);

			Print(node->GetLeftNode(), Indent, Builder);
			PrintLine(node->GetOperatorName(), Indent, Builder);
			Print(node->GetRightNode(), Indent, Builder);
		}
		else if (IS_CHILD_OF(Node, ArithmeticOperatorNode))
		{
			ArithmeticOperatorNode *node = reinterpret_cast<ArithmeticOperatorNode*>(Node);

			Print(node->GetLeftNode(), Indent, Builder);
			PrintLine(node->GetOperatorName(), Indent, Builder);
			Print(node->GetRightNode(), Indent, Builder);
		}
		else
			throw std::exception("Unknown TreeNode type");
	}

	void Expression::PrintLine(const std::string &Text, int Indent, std::stringstream &Builder)
	{
		for (int i = 0; i < Indent; ++i)
			Builder << ' ';

		Builder << Text << std::endl;
	}
}